### ISSUE TITLE
Use RuntimeGeneratedFunctions to eliminate the use of eval without world-age problems

### ISSUE THREAD

**Issue by ChrisRackauckas** [cafaxo/FrechetDiff.jl#1]
Explanatory title

**Comment by cafaxo** [comment]
As far as I understand, closures are not supported in generated functions.
However, the current design of this package heavily relies on closures within the generated code.

In order to really get rid of `eval` / `invokelatest`, a major redesign would probably be necessary.

For me, the purpose of creating this package was to
 1. gain a better understanding of AD with Julia in general
 2. try to see if there is a different approach that works better for some special problems
 
If people are actually interested in this approach (I would be glad to discuss this package in general), I would consider restarting its development.

As a workaround for now, one could use the following pattern to prevent performance problems:
```julia
# function to be differentiated
function f(x)
  # do something with x
end

# expensive function that uses derivative of f multiple times (like an ODE solver)
function g(fd)
  # use derivative fd of f
end

function solve_problem(f)
  # compute derivative fd via FrechetDiff
  return Base.invokelatest(g, fd)
end
```

**Comment by ChrisRackauckas** [comment]
>As far as I understand, closures are not supported in generated functions.

That is true, but RuntimeGeneratedFunctions can be used in RuntimeGeneratedFunctions and that could be how to handle it.

Though I think in general, you don't want to be doing this in the Julia runtime but instead as a compiler plugin.
